### PR TITLE
[Reviewer: Ellie, Rob] Chronos refactor and introduction of SNMP handler to Chronos

### PR DESCRIPTION
**Pull request by TygerBurning**
Restructure of Chronos removes most logic from the TimerStore and puts it in the TimerHandler. This requires changing a lot of UTs.

Deleting timers is now solely the responsibility of the Timer Handler, as is reporting statistics.

### CODE REVIEW

File: modules/cpp-common [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit fc9944f163203b746269d2f3c7282d425b3fcea5
+Subproject commit 061699fefd8cffa52842b1dffd7fa48406e6d13a
```
Comment (eleanor-merry) [inline modules/cpp-common]: You'll need to update this when you merge your cpp-common changes

File: scripts/chronos_resync.py [hunk 2]
```diff
@@ -232,32 +231,35 @@ def write_scale_down_config_for_all_nodes(self, leaving_lower, leaving_upper):
...
         # adds 100 timers, scales up to 4 Chronos nodes, then checks that
         # 100 timers pop.
- 
+
         # Start initial nodes and add timers
         self.write_config_for_nodes(0, 2)
         start_nodes(0, 2)
         create_timers(chronos_nodes[0], 100)
 
+
```
Comment (eleanor-merry) [inline scripts/chronos_resync.py]: Extra lines?

File: scripts/chronos_resync.py [hunk 3]
```diff
@@ -146,7 +146,6 @@ def create_timers(target, num):
             }
         }
     }
-
```
Comment (rkd-msw) [inline scripts/chronos_resync.py]: This line should probably stay.

File: scripts/chronos_resync.py (hunk 2 above)
Comment (rkd-msw) [inline scripts/chronos_resync.py]: Have you run flake8 over this Python and checked the warnings?

File: src/include/timer_store.h [hunk 4]
```diff
@@ -45,33 +45,84 @@
...
+    return false;
+  }
+
+  bool operator<(const TimerPair &other) const
+  {
+    return (active_timer->id < other.active_timer->id);
+  }
+};
+
+namespace std
```
Comment (rkd-msw) [inline src/include/timer_store.h]: I hadn't seen this fomulation before - reading http://stackoverflow.com/questions/8157937/how-to-specialize-stdhashkeyoperator-for-user-defined-type-in-unordered I think it's right, but worth a comment to explain what you're trying to achieve.

File: src/main/timer_handler.cpp [hunk 5]
```diff
@@ -230,19 +516,59 @@ void TimerHandler::pop(Timer* timer)
...
-  if (_current_timers_scalar != NULL)
+}
+
+bool TimerHandler::overflow_less_than(uint32_t a, uint32_t b)
+{
+  return ((a - b) > ((uint32_t)(1) << 31));
+}
+
+// Report an update to the number of timers to statistics
+void TimerHandler::update_statistics(std::vector<std::string> new_tags,
```
Comment (rkd-msw) [inline src/main/timer_handler.cpp]: How well does this perform? I'm nervous that this is a big chunk of extra processing/copying on every timer - can we optimise this at all?

For example:
* passing the arguments by reference to eliminate copying
* parsing tags into a std::set when we first get them (so we don't need to copy them into one here)
* somehow optimising the common cases - completely new timers or timers with unchanged tags?

File: src/main/timer_handler.cpp [hunk 6]
```diff
@@ -230,19 +516,59 @@ void TimerHandler::pop(Timer* timer)
...
+                                     std::vector<std::string> old_tags)
```
Comment (rkd-msw) [inline src/main/timer_handler.cpp]: Prototype is `void update_statistics(std::vector<std::string> new_tags, std::vector<std::string> existing_tags);` - consistent naming?

File: src/main/timer_handler.cpp [hunk 7]
```diff
@@ -88,31 +91,200 @@ TimerHandler::~TimerHandler()
...
+
+  _store->insert(new_tp, id, next_pop_time, cluster_view_id_vector);
+}
+
+void TimerHandler::return_timer_to_store(Timer* timer, bool successful)
+{
+  if (!successful)
+  {
+    // In this case, we should update our local statistics, and set the alarm
+    update_statistics(std::vector<std::string>(), timer->tags);
```
Comment (rkd-msw) [inline src/main/timer_handler.cpp]: Why are we clearing all the tags here?

File: src/main/timer_handler.cpp [hunk 8]
```diff
@@ -230,19 +516,59 @@ void TimerHandler::pop(Timer* timer)
...
   }
-  if (_current_timers_scalar != NULL)
+}
+
+bool TimerHandler::overflow_less_than(uint32_t a, uint32_t b)
+{
+  return ((a - b) > ((uint32_t)(1) << 31));
+}
+
+// Report an update to the number of timers to statistics
```
Comment (rkd-msw) [inline src/main/timer_handler.cpp]: Worth explaining the usage in the comment - in particular, I think there's a need to call this with an empty set of tags if the timer is deleted?

File: src/include/timer_store.h [hunk 9]
```diff
@@ -45,33 +45,84 @@
...
-
-  // Get timer information from the store for timers where
-  // request_node should be a replica
-  virtual HTTPCode get_timers_for_node(std::string request_node,
-                                       int max_responses,
-                                       std::string cluster_view_id,
-                                       std::string& get_response);
+  // Fetch timers with a view ID different to the one given,
+  // and populate the vector with 1000 relevant timers
+  virtual bool get_by_not_view_id(std::string cluster_view_id, int max_responses, std::unordered_set<TimerPair>&);
```
Comment (rkd-msw) [inline src/include/timer_store.h]: Could use a better name.

File: src/include/timer_store.h [hunk 10]
```diff
@@ -45,33 +45,84 @@
 #include <map>
 #include <string>
 
+struct TimerPair {
```
Comment (rkd-msw) [inline src/include/timer_store.h]: Worth a class comment to explain what this is for?

**Comment by rkd-msw** [comment]
I've reviewed the SNMP stuff and a bit of non-SNMP stuff that jumped out at me.

File: src/include/timer_handler.h [hunk 11]
```diff
@@ -47,15 +47,18 @@
...
 #include "snmp_scalar.h"
 
 class TimerHandler
 {
 public:
-  TimerHandler(TimerStore*, Callback*, SNMP::ContinuousAccumulatorTable*, SNMP::U32Scalar*);
+  TimerHandler(TimerStore*, Callback*, Replicator*, Alarm*, SNMP::InfiniteTimerCountTable*);
   virtual ~TimerHandler();
-  virtual void add_timer(Timer*);
+  virtual void add_timer_to_store(Timer*);
```
Comment (eleanor-merry) [inline src/include/timer_handler.h]: Not sure about the naming here - users of timer handler don't need to know what it's doing with it.

File: src/include/timer_handler.h [hunk 12]
```diff
@@ -71,18 +74,23 @@ class TimerHandler
...
 
 private:
-  void pop(std::unordered_set<Timer*>&);
+  void pop(std::unordered_set<TimerPair>&);
   void pop(Timer*);
-  void signal_new_timer(unsigned int);
-  void update_statistics(uint32_t);
+  bool timer_is_on_node(std::string, std::string, Timer*, std::vector<std::string>&);
+  void set_tombstone_values(Timer* timer, Timer* existing);
+  bool overflow_less_than(uint32_t a, uint32_t b);
```
Comment (eleanor-merry) [inline src/include/timer_handler.h]: This looks like a useful function - can you put it in utils in cpp-common instead?

File: src/test/test_timer_store.cpp [hunk 13]
```diff
@@ -142,22 +142,50 @@ class TestTimerStore : public Base
...
+    delete timers[2]; timers[2] = NULL;
+    delete tombstone; tombstone = NULL;
+
     delete ts; ts = NULL;
     delete hc; hc = NULL;
     cwtest_reset_time();
     Base::TearDown();
   }
 
+  void ts_insert_helper(TimerPair pair)
```
Comment (rkd-msw) [inline src/test/test_timer_store.cpp]: I assumed on first glance that this would insert a helper - is there a better name?

File: src/include/timer_store.h [hunk 14]
```diff
@@ -45,33 +45,84 @@
...
 {
 public:
+
   TimerStore(HealthChecker* hc);
   virtual ~TimerStore();
 
-  // Add a timer to the store.
-  virtual void add_timer(Timer*);
+  // Insert a timer (with an ID that doesn't exist already)
+  virtual void insert(TimerPair, TimerID, uint32_t, std::vector<std::string>);
```
Comment (eleanor-merry) [inline src/include/timer_store.h]: Can you put names for your parameters in the prototype? It's obvious what TimerPair/TimerID are, but not at all what the uint32_t/std::vector are used for

File: src/include/timer_store.h [hunk 15]
```diff
@@ -81,9 +132,9 @@ class TimerStore
   // - A set of overdue timers.
   //
   // New timers are placed into on of these structures:
-  // - The short wheel if due to pop in the next second.
-  // - The long wheel if due to pop in the next hour (but not the next second).
-  // - The heap if due to pop >=1hr in the future.
+  // - The short wheel if due to pop in ~the next second.
```
Comment (eleanor-merry) [inline src/include/timer_store.h]: ~? Can you be clearer when a timer will be placed into a particular wheel?

File: src/main/http_callback.cpp [hunk 16]
```diff
@@ -153,12 +139,7 @@ void HTTPCallback::worker_thread_entry_point()
                   timer->callback_url.c_str(),
                   curl_easy_strerror(curl_rc));
 
-      if (_timer_pop_alarm && timer->is_last_replica())
-      {
-        _timer_pop_alarm->set();
-      }
-
-      delete timer;
+      _handler->return_timer_to_store(timer, false);
```
Comment (eleanor-merry) [inline src/main/http_callback.cpp]: Set timer = NULL here?

File: src/main/timer_handler.cpp [hunk 17]
```diff
@@ -88,31 +91,200 @@ TimerHandler::~TimerHandler()
...
+      existing_tp.information_timer = NULL;
+
+      delete timer;
+    }
+    else
+    {
+      // We want to add the timer, so decide whether this is an update, or
+      // if we need to save off the old timer (as an information timer)
+
+      set_tombstone_values(timer, existing_tp.active_timer);
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: I think if we're moving the active timer to an informational timer we don't want to tombstone it

Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: But we do want to update the replica tracker values

File: src/main/timer_handler.cpp [hunk 18]
```diff
@@ -88,31 +91,200 @@ TimerHandler::~TimerHandler()
...
+        new_tp.active_timer = timer;
+        if (existing_tp.information_timer)
+        {
+          new_tp.information_timer = existing_tp.information_timer;
+          existing_tp.information_timer = NULL;
+        }
+      }
+    }
+  }
+  // Update statistics
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: Add a line break here please

File: src/main/timer_handler.cpp [hunk 19]
```diff
@@ -122,13 +294,123 @@ HTTPCode TimerHandler::get_timers_for_node(std::string request_node,
                                            std::string& get_response)
 {
   pthread_mutex_lock(&_mutex);
-  HTTPCode rc = _store->get_timers_for_node(request_node,
-                                            max_responses,
-                                            cluster_view_id,
-                                            get_response);
+  std::unordered_set<TimerPair> timers;
+  bool finished = _store->get_by_not_view_id(cluster_view_id, max_responses, timers);
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: This doesn't looks quite right - I think you can end up finishing the resync too early. I'll come and talk to you about this

File: src/main/timer_handler.cpp [hunk 20]
```diff
@@ -138,20 +420,20 @@ HTTPCode TimerHandler::get_timers_for_node(std::string request_node,
...
   std::unordered_set<Timer*>::iterator sample_timer;
 
   pthread_mutex_lock(&_mutex);
 
-  _store->get_next_timers(next_timers);
+  _store->fetch_next_timers(next_timers);
 
   while (!_terminate)
   {
+
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: Extra line?

File: src/main/timer_handler.cpp [hunk 21]
```diff
@@ -161,13 +443,13 @@ void TimerHandler::run() {
       struct timespec next_pop;
       clock_gettime(CLOCK_MONOTONIC, &next_pop);
 
-      if (next_pop.tv_nsec < 990 * 1000 * 1000)
+      if (next_pop.tv_nsec < 790 * 1000 * 1000)
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: Can you comment why it's 790?

File: src/main/timer_handler.cpp (hunk 17 above)
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: Ok, I understand what the set_tombstone function is doing now. Can you change the name to copy/save_tombstone to make it clearer?

File: src/main/timer_handler.cpp (hunk 5 above)
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: +1 on @rkd-msw comments

File: src/main/timer_store.cpp [hunk 22]
```diff
@@ -55,13 +55,13 @@
...
-#define TIMER_LOG_PARAMS(T) (T)->id,                                           \
-                            (T)->start_time_mono_ms,                           \
-                            (T)->interval_ms,                                     \
-                            (T)->repeat_for,                                   \
-                            (T)->sequence_number,                              \
-                            (T)->callback_url.c_str(),                         \
-                            (T)->callback_body.c_str()
+#define TIMER_LOG_PARAMS(T) (T).active_timer->id,                                           \
+                            (T).active_timer->start_time_mono_ms,                           \
+                            (T).active_timer->interval_ms,                                     \
```
Comment (eleanor-merry) [inline src/main/timer_store.cpp]: \ alignment

File: src/include/timer_store.h (hunk 14 above)
Comment (eleanor-merry) [inline src/include/timer_store.h]: It's a logic error to insert with an existing TimerID - how is this enforced?

File: src/main/timer_store.cpp [hunk 23]
```diff
@@ -229,100 +138,56 @@ void TimerStore::add_timer(Timer* t)
...
                 "LONG_WHEEL_NUM_BUCKETS constant");
-    _extra_heap.push_back(t);
+    _extra_heap.push_back(tp);
     std::push_heap(_extra_heap.begin(), _extra_heap.end());
   }
 
+  // Add to the view specific mapping for easy retrieval on resync
+  for (std::vector<std::string>::iterator it = cluster_view_id_vector.begin();
+                                          it != cluster_view_id_vector.end();
+                                             ++it)
```
Comment (eleanor-merry) [inline src/main/timer_store.cpp]: Alignment

File: src/main/timer_store.cpp [hunk 24]
```diff
@@ -229,100 +138,56 @@ void TimerStore::add_timer(Timer* t)
...
+    _timer_view_id_table[timer.active_timer->cluster_view_id].erase(timer.active_timer->id);
 
-          // LCOV_EXCL_START
-          TRC_ERROR("Failed to remove timer consistently");
-          purge_timer_from_wheels(timer);
-          // LCOV_EXCL_STOP
-        }
-      }
+    // Remove from cluster structure
+    if (timer.information_timer) {
```
Comment (eleanor-merry) [inline src/main/timer_store.cpp]: Hadouken style braces please

File: src/main/timer_store.cpp [hunk 25]
```diff
@@ -346,6 +211,60 @@ void TimerStore::get_next_timers(std::unordered_set<Timer*>& set)
...
+
+  for (std::map<std::string, std::unordered_set<TimerID>>::iterator it = _timer_view_id_table.begin();
+                                                              it != _timer_view_id_table.end();
+                                                              ++it)
+  {
+    if (it->first != new_cluster_view_id)
+    {
+      // We've found another set of interesting cluster view ids
+      done = false;
+      if ((max_responses - view_ids.size() > it->second.size()))
```
Comment (eleanor-merry) [inline src/main/timer_store.cpp]: It doesn't look like your brackets are quite right here

File: src/main/timer_store.cpp [hunk 26]
```diff
@@ -490,226 +421,77 @@ void TimerStore::refill_short_wheel()
...
-{
-  if (t->is_tombstone())
-  {
-    // Learn the interval so that this tombstone lasts long enough to catch
-    // errors.
-    t->interval_ms = existing->interval_ms;
-    t->repeat_for = existing->interval_ms;
-  }
 }
+// LCOV_EXCL_STOP
```
Comment (eleanor-merry) [inline src/main/timer_store.cpp]: There's an overflow_less_than function just below this line - hasn't this moved to the timer handler?

File: src/test/fakesnmp.cpp [hunk 27]
```diff
@@ -0,0 +1,57 @@
+/**
```
Comment (eleanor-merry) [inline src/test/fakesnmp.cpp]: Can these files have more specific names?

File: src/test/test_timer_handler.cpp [hunk 28]
```diff
@@ -248,34 +268,422 @@ TEST_F(TestTimerHandler, AddTimer)
...
+
+  cwtest_reset_time();
+
+  // timer1 is deleted by handler
+  delete timer2;
+}
+
+// Test that if there is already an information timer for this timer
+// we should give a warning that we are replacing this timer
+TEST_F(TestTimerHandler, AddExistingTimerPair)
```
Comment (eleanor-merry) [inline src/test/test_timer_handler.cpp]: Is this comment up to date? Can you make sure that for each test you have a short comment saying what the test is testing?

File: src/test/test_timer_handler.cpp [hunk 29]
```diff
@@ -248,34 +268,422 @@ TEST_F(TestTimerHandler, AddTimer)
...
+
+  EXPECT_CALL(*_store, fetch(timer2->id, _)).
+                       WillOnce(DoAll(SetArgReferee<1>(return_pair1),Return(true)));
+  EXPECT_CALL(*_store, insert(_, timer2->id, timer2->next_pop_time(), _)).
+              WillOnce(SaveArg<0>(&insert_pair));
+
+  _th->add_timer_to_store(timer2);
+
+  EXPECT_EQ(insert_pair.active_timer, timer2);
+//  EXPECT_EQ(insert_pair.information_timer, timer1);
```
Comment (eleanor-merry) [inline src/test/test_timer_handler.cpp]: Commented out code?

File: src/test/test_timer_store.cpp [hunk 30]
```diff
@@ -901,112 +860,149 @@ TYPED_TEST(TestTimerStore, SelectTimersNoMatchesClusterID)
...
+
+  // Get the timers out of the store to clean up properly
+  std::unordered_set<TimerPair> next_timers;
+  cwtest_advance_time_ms(TestFixture::timers[0]->interval_ms +
+                         TestFixture::timers[1]->interval_ms +
+                         TestFixture::timers[2]->interval_ms + TIMER_GRANULARITY_MS);
+  TestFixture::ts->fetch_next_timers(next_timers);
 }
 
+/*TYPED_TEST(TestTimerStore, ModifySavedTimers)
```
Comment (eleanor-merry) [inline src/test/test_timer_store.cpp]: Commented out test?

**Comment by eleanor-merry** [comment]
I've reviewed the Chronos refactor. Generally looks good to me, but I think there are a few edge cases on resync that you haven't covered. I'll come and talk to you about these

File: src/main/timer_handler.cpp [hunk 31]
```diff
@@ -88,31 +94,207 @@ TimerHandler::~TimerHandler()
...
+      int remaining_replicas = timer->update_replica_tracker(replica_index);
+
+      if (remaining_replicas == 0)
+      {
+        if (!timer_in_wheel)
+        {
+          // All the new replicas have been told about the timer. We don't
+          // need to store the information about the timer anymore.
+          delete timer; timer = NULL;
+          store_timers.information_timer = NULL;
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: This line isn't needed (but check this in UTs)

File: src/main/timer_handler.cpp [hunk 32]
```diff
@@ -122,13 +304,134 @@ HTTPCode TimerHandler::get_timers_for_node(std::string request_node,
...
+
+  int retrieved_timers = 0;
+
+  for (TimerStore::TSIterator it = _store->begin(cluster_view_id);
+       it != _store->end();
+       ++it)
+  {
+    Timer* timer_copy;
+    TimerPair pair = *it;
+    if (!pair.information_timer)
```
Comment (eleanor-merry) [inline src/main/timer_handler.cpp]: You should look at the active timer first, and use that if the cluster view ID is out of date
